### PR TITLE
Add “doctor” chezmoi command

### PR DESCRIPTION
**Pull request by kkirsche**
Add a `chezmoi doctor` command. Not sure the best way to test this though (unit test wise). Thoughts?

```
 ~/git/go/src/github.com/twpayne/chezmoi >> feature/cmd_doctor ? $ go build -o cmoi                                 ✔
 ~/git/go/src/github.com/twpayne/chezmoi >> feature/cmd_doctor ? $ ./cmoi doctor
Arch: amd64
OS: darwin
Configuration File: Exists (-rw-r--r-- /Users/kkirsche/.config/chezmoi/chezmoi.yaml)
Destination Directory: Exists (-rwxr-xr-x /Users/kkirsche)
Source Directory: Exists (-rwx------ /Users/kkirsche/.local/share/chezmoi)
Hashicorp Vault: Installed (-r-xr-xr-x /Users/kkirsche/.local/share/chezmoi)
Bitwarden: Not Found
LastPass: Not Found
```

### CODE REVIEW

**Comment by kkirsche** [comment]
This PR is related to the `doctor` command mentioned in #134

File: cmd/doctor.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+func init() {
+	rootCommand.AddCommand(doctorCommand)
+
+}
+
+func (c *Config) runDoctorCommand(fs vfs.FS, args []string) error {
+	fmt.Printf("Arch: %s\n", runtime.GOARCH)
+	fmt.Printf("OS: %s\n", runtime.GOOS)
+
+	info, err := os.Stat(configFile)
```
Comment (twpayne) [inline cmd/doctor.go]: Please use `fs.Stat` here and below so this command can be tested using [`vfst`](https://godoc.org/github.com/twpayne/go-vfs/vfst) eventually.

File: cmd/doctor.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	if err == nil {
+		info, err = os.Stat(path)
+		if !os.IsNotExist(err) {
+			fmt.Printf("LastPass: Installed (%s %s)\n", info.Mode().Perm().String(), path)
+		} else {
+			fmt.Printf("LastPass: Error (%s)\n", err.Error())
+		}
+	} else {
+		fmt.Println("LastPass: Not Found")
+	}
```
Comment (twpayne) [inline cmd/doctor.go]: Please refactor these three tests into a `for`-loop over a slice, something like:

```go
    for _, binary := range []struct{
        description string
        name        string
    }{
        {description: "Bitwarden", name: config.Bitwarden.BW},
        {description: "LastPass", name: config.LastPass.Lpass},
        {description: "Vault", name: config.Vault.Vault},
    } {
        if path, err := exec.LookPath(binary.name); err == nil {
            // ...
        } else {
            // ...
        }
    }
```

File: cmd/doctor.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	} else {
+		fmt.Printf("Destination Directory: Exists (%s %s)\n", info.Mode().Perm().String(), config.DestDir)
+	}
+
+	info, err = os.Stat(config.SourceDir)
+	if os.IsNotExist(err) {
+		fmt.Printf("Source Directory: Does Not Exist (%s)\n", config.SourceDir)
+	} else {
+		fmt.Printf("Source Directory: Exists (%s %s)\n", info.Mode().Perm().String(), config.SourceDir)
+	}
```
Comment (twpayne) [inline cmd/doctor.go]: Please refactor the `config.DestDir` and `config.SourceDir` tests into a `for`-loop over a slice, as described for the binaries.

**Comment by kkirsche** [comment]
If you prefer aligned, we can use a tabwriter (https://golang.org/pkg/text/tabwriter/) to align it like so:

```
Arch:                    amd64
OS:                      darwin
Configuration File:      Exists      (-rw-r--r-- /Users/kkirsche/.config/chezmoi/chezmoi.yaml)
Destination Directory:   Exists      (-rwxr-xr-x /Users/kkirsche)
Source Directory:        Exists      (-rwx------ /Users/kkirsche/.local/share/chezmoi)
Hashicorp Vault:         Installed   (-r-xr-xr-x /usr/local/bin/vault)
Bitwarden:               Not Found
LastPass:                Not Found
```

**Comment by twpayne** [comment]
#138 is also relevant now :)

File: cmd/doctor.go (hunk 3 above)
Comment (kkirsche) [reply cmd/doctor.go]: Updated

File: cmd/doctor.go (hunk 2 above)
Comment (kkirsche) [reply cmd/doctor.go]: Updated

**Comment by kkirsche** [comment]
Output:

```
 ~/git/go/src/github.com/twpayne/chezmoi >> feature/cmd_doctor ? $ ./cmoi doctor
Arch:                    amd64
OS:                      darwin
Configuration File:      Exists   (-rw-r--r-- /Users/kkirsche/.config/chezmoi/chezmoi.yaml)
Destination Directory:   Exists   (-rwxr-xr-x /Users/kkirsche)
Source Directory:        Exists   (-rwx------ /Users/kkirsche/.local/share/chezmoi)
Bitwarden:               Not Found
LastPass:                Not Found
Vault:                   Installed   (-r-xr-xr-x /usr/local/bin/vault)
```

**Comment by twpayne** [comment]
For the output, let's focus on the logic first and then make the UX pretty second :)

File: cmd/doctor.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	"os/exec"
+	"runtime"
+
+	"github.com/spf13/cobra"
+	vfs "github.com/twpayne/go-vfs"
+)
+
+// doctorCmd represents the doctor command
+var doctorCommand = &cobra.Command{
+	Use:   "doctor",
```
Comment (twpayne) [inline cmd/doctor.go]: This also needs `Args: cobra.NoArgs`.

File: cmd/doctor.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+* Is Hashicorp Vault installed?
+* Is Bitwarden installed?
+* Is LastPass installed?
+`,
+	RunE: makeRunE(config.runDoctorCommand),
+}
+
+func init() {
+	rootCommand.AddCommand(doctorCommand)
+
```
Comment (twpayne) [inline cmd/doctor.go]: nit: remove blank line

File: cmd/doctor.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	for _, dir := range []struct {
+		description string
+		path        string
+	}{
+		{description: "Configuration File", path: configFile},
+		{description: "Destination Directory", path: config.DestDir},
+		{description: "Source Directory", path: config.SourceDir},
+	} {
+		info, err := os.Stat(dir.path)
```
Comment (twpayne) [inline cmd/doctor.go]: Please use `fs.Stat`, not `os.Stat`. Also, I don't think `info` is used. You can write this as:
```go
    if _, err := fs.Stat(dir.path); err == nil {
        // print ok message
    } else {
        // print not ok message
    }
```

File: cmd/doctor.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+		if err == nil {
+			info, err := os.Stat(path)
+			if !os.IsNotExist(err) {
+				fmt.Fprintf(w, "%s:\tInstalled\t(%s %s)\n", binary.description, info.Mode().Perm().String(), path)
+			} else {
+				fmt.Fprintf(w, "%s:\tError\t(%s)\n", binary.description, err.Error())
+			}
+		} else {
+			fmt.Fprintf(w, "%s:\tNot Found\n", binary.description)
+		}
```
Comment (twpayne) [inline cmd/doctor.go]: Please use the structure requested for checking directories and files above.

File: cmd/doctor.go (hunk 5 above)
Comment (kkirsche) [reply cmd/doctor.go]: Updated

File: cmd/doctor.go (hunk 4 above)
Comment (kkirsche) [reply cmd/doctor.go]: Updated

**Comment by kkirsche** [comment]
Updated output:

```
 ~/git/go/src/github.com/twpayne/chezmoi >> feature/cmd_doctor ? $ go build -o cmoi
 ~/git/go/src/github.com/twpayne/chezmoi >> feature/cmd_doctor ? $ ./cmoi doctor
Arch:                    amd64
OS:                      darwin
Configuration File:      Exists   (-rw-r--r-- /Users/kkirsche/.config/chezmoi/chezmoi.yaml)
Destination Directory:   Exists   (-rwxr-xr-x /Users/kkirsche)
Source Directory:        Exists   (-rwx------ /Users/kkirsche/.local/share/chezmoi)
Bitwarden:               Not Found
LastPass:                Not Found
Pass:                    Not Found
Vault:                   Installed   (-r-xr-xr-x /usr/local/bin/vault)
```

**Comment by twpayne** [comment]
Another thing to add: please return `nil` if everything is ok, and call `os.Exit(1)` if it is not. This will allow `chezmoi doctor` to be used in scripts.

File: cmd/doctor.go (hunk 1 above)
Comment (kkirsche) [reply cmd/doctor.go]: Updated

**Comment by kkirsche** [comment]
What situations do you define as "not ok"?

File: cmd/doctor.go (hunk 6 above)
Comment (kkirsche) [reply cmd/doctor.go]: info is used in all cases to print permissions if it exists

File: cmd/doctor.go (hunk 7 above)
Comment (kkirsche) [reply cmd/doctor.go]: I don't believe this is applicable, as described above info is used, and thus is required to be used. If it is, more specificity in the request would be appreciated

**Comment by twpayne** [comment]
> What situations do you define as "not ok"?

Good question! I would say anything that prevents `chezmoi` from running at all, e.g. missing source or destination directories. Missing config files is OK, and missing password manager binaries is OK.

**Comment by kkirsche** [comment]
Added exit codes 👍

**Comment by kkirsche** [comment]
Updated output:

```
Arch:                    amd64
OS:                      darwin
Configuration File:      Exists      (-rw-r--r-- /Users/kkirsche/.config/chezmoi/chezmoi.yaml)
Destination Directory:   Exists      (-rwxr-xr-x /Users/kkirsche)
Source Directory:        Exists      (-rwx------ /Users/kkirsche/.local/share/chezmoi)
Bitwarden:               Not Found   (looking for: bw)
LastPass:                Not Found   (looking for: lpass)
Pass:                    Not Found   (looking for: pass)
Vault:                   Installed   (-r-xr-xr-x /usr/local/bin/vault)
VCS:                     Installed   (-rwxr-xr-x /usr/local/bin/git)
```

**Comment by twpayne** [comment]
Thanks for addressing the commands. Please can you squash the commits in to a single one with the message "Add doctor command" and I will then merge.

**Comment by kkirsche** [comment]
Well, squashing stuff broke the PR 😢

**Pull request closed without merging**